### PR TITLE
Home task 3

### PR DESCRIPTION
**Pull request by oksanazah**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,43 +1,47 @@
...
-import { mockedAuthorsList, mockedCoursesList } from './constants';
-import type { Course, Author } from './helpers/interfaces';
+import { getCourses } from './store/courses/actionCreators';
+import { getAuthors } from './store/authors/actionCreators';
+import { selectUser } from './helpers';
+import type { User } from './helpers';
 
 import './App.css';
 
+const store = createStore(rootReducer, composeWithDevTools());
```
Comment (KitTheProgrammer) [inline src/App.tsx]: this line should be in `src/store/index.ts`

File: src/App.tsx [hunk 2]
```diff
@@ -1,43 +1,47 @@
...
+type RootState = ReturnType<typeof store.getState>;
+
 const App: React.FC = () => {
-	const [courseList, setCourseList] = useState<Course[]>(mockedCoursesList);
-	const [authorList, setAuthorList] = useState<Author[]>(mockedAuthorsList);
-	const [userName, setUserName] = useState<string | null>('');
+	const dispatch = useDispatch();
+	const userName: User = useSelector(selectUser);
```
Comment (KitTheProgrammer) [inline src/App.tsx]: I think type inferring can be applied here if you configured your selector correctly

File: src/App.tsx [hunk 3]
```diff
@@ -1,43 +1,47 @@
...
 
 	useEffect(() => {
-		setUserName(localStorage.getItem('name'));
-	}, []);
-
-	const newAuthorList = (authorList: Author[]): void => {
-		setAuthorList(authorList);
-	};
+		getCourses().then((data) => dispatch(data));
+	}, [dispatch]);
```
Comment (KitTheProgrammer) [inline src/App.tsx]: It's not that important, but callback inside the useEffect can be async
`useEffect(async () => {})`

File: src/components/CourseInfo/CourseInfo.tsx [hunk 4]
```diff
@@ -1,16 +1,25 @@
...
+
+	useEffect(() => {
+		if (courseId) {
+			getCourseInfo(courseId).then((data: CourseInfoResponse): void => {
+				if (data.successful) {
+					setCourse(data.result);
+				}
+			});
+		}
+	}, [courseId]);
```
Comment (KitTheProgrammer) [inline src/components/CourseInfo/CourseInfo.tsx]: using async here will be beneficial

File: src/components/Header/Header.tsx [hunk 5]
```diff
@@ -1,20 +1,24 @@
...
 
 const Header: React.FC<HeaderProps> = ({ userName }) => {
 	const location = useLocation();
 	const navigate = useNavigate();
+	const dispatch = useDispatch();
 
 	const logout = (): void => {
+		dispatch(onLogout());
 		localStorage.removeItem('token');
 		localStorage.removeItem('name');
```
Comment (KitTheProgrammer) [inline src/components/Header/Header.tsx]: You can remove items from local storage from your `onLogout` action

File: src/components/Login/Login.tsx [hunk 6]
```diff
@@ -39,12 +42,15 @@ const Login: React.FC = () => {
 	): Promise<void> => {
 		e.preventDefault();
 
-		const result = await auth(user, 'login');
+		const result = await onLogin(user, 'login');
 
-		localStorage.setItem('token', `${result?.result}`);
-		localStorage.setItem('name', `${result?.user?.name}`);
+		localStorage.setItem('token', `${result.payload?.result}`);
+		localStorage.setItem('name', `${result.payload?.user?.name}`);
```
Comment (KitTheProgrammer) [inline src/components/Login/Login.tsx]: local storage operations can be moved to your `onLogin` action

File: src/helpers/selectors.tsx
Comment (KitTheProgrammer) [inline src/helpers/selectors.tsx:1]: this file have to be in `src/store` or `src/store/selectors`

File: src/services.tsx [hunk 7]
```diff
@@ -31,4 +33,25 @@ const auth = async (user: User, url: string): Promise<AuthPops | undefined> => {
...
+
+	return courseInfo;
+};
+
+const getAuthorsList = async (): Promise<AuthorResponse> => {
+	const response: Response = await fetch('http://localhost:4000/authors/all');
+	const authorsList: Promise<AuthorResponse> = response.json();
+
+	return authorsList;
+};
```
Comment (KitTheProgrammer) [inline src/services.tsx]: can't see any error handling here.

if you have it at places you are using these functions then it is better to move them here

File: src/store/authors/actionCreators.tsx [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+export const getAuthors = (): Promise<AuthorAction> => {
+	const res = getAuthorsList().then(
+		(data: AuthorResponse): AuthorAction => ({
+			type: GET_AUTHORS,
+			payload: data.result,
+		})
+	);
+
+	return res;
+};
```
Comment (KitTheProgrammer) [inline src/store/authors/actionCreators.tsx:19]: no error handling

File: src/store/courses/actionCreators.tsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+export const getCourses = (): Promise<CourseAction> => {
+	const res = getCoursesList().then(
+		(data: CourseResponse): CourseAction => ({
+			type: GET_COURSES,
+			payload: data.result,
+		})
+	);
+
+	return res;
+};
```
Comment (KitTheProgrammer) [inline src/store/courses/actionCreators.tsx:24]: no error handling

**Review by KitTheProgrammer — CHANGES_REQUESTED** [review]
No bugs, good job)

**Review by EpamFanAnton — APPROVED** [review]
good job 15

**Pull request merged**